### ISSUE TITLE
Classgraph finding duplicate resources with relative classpath elements

### ISSUE THREAD

**Issue by facboy** [classgraph/classgraph#297]
Classgraph seems to pull classpath elements both from the ClassLoaders and from the `java.class.path` system property.  When the classpath contains jars with absolute paths that contain relative elements, then they show up in the ClassLoaders with their canonical path, but remain as absolute-relative paths in `java.class.path`.  Classgraph does not detect this and so returns duplicate entries when there are not actually any duplicate entries in the 'real' classpath.

https://github.com/facboy/classgraph-duplicate-bug

**Comment by lukehutch** [comment]
Thanks for the report. There was a big change to how paths were handled in 4.6.0, and it looks like the path canonicalization step was accidentally dropped. I'll take a look at this soon (I'm currently traveling).

**Comment by lukehutch** [comment]
@facboy your project doesn't generate duplicates on Linux. Are you running on Windows? Can you please attach your verbose logs?

**Comment by facboy** [comment]
Are you running `mvn clean verify` or are you running it in an IDE?

```
Jan 08, 2019 9:14:55 PM nonapi.io.github.classgraph.utils.LogNode flush
INFO: 2019-01-08T21:14:55.465+0000	ClassGraph	ClassGraph version 4.6.12
2019-01-08T21:14:55.466+0000	ClassGraph	Operating system: Linux 3.10.0-957.1.3.el7.x86_64 amd64
2019-01-08T21:14:55.466+0000	ClassGraph	Java version: 1.8.0_191 / 1.8.0_191-b12 (Oracle Corporation)
2019-01-08T21:14:55.466+0000	ClassGraph	Java home: /usr/lib/jvm/java-1.8.0-openjdk-1.8.0.191.b12-1.el7_6.x86_64/jre
2019-01-08T21:14:55.472+0000	ClassGraph	JRE rt.jar:
2019-01-08T21:14:55.472+0000	ClassGraph	-- /usr/lib/jvm/java-1.8.0-openjdk-1.8.0.191.b12-1.el7_6.x86_64/jre/lib/rt.jar
2019-01-08T21:14:55.478+0000	ClassGraph	ScanSpec:
2019-01-08T21:14:55.478+0000	ClassGraph	-- packageWhiteBlackList: whitelist: [META-INF.maven.io.github.classgraph.classgraph.]
2019-01-08T21:14:55.478+0000	ClassGraph	-- packagePrefixWhiteBlackList: whitelistPrefixes: [META-INF.maven.io.github.classgraph.classgraph..]
2019-01-08T21:14:55.478+0000	ClassGraph	-- pathWhiteBlackList: whitelist: [META-INF/maven/io.github.classgraph/classgraph/]
2019-01-08T21:14:55.478+0000	ClassGraph	-- pathPrefixWhiteBlackList: whitelistPrefixes: [META-INF/maven/io.github.classgraph/classgraph/]
2019-01-08T21:14:55.479+0000	ClassGraph	-- classWhiteBlackList: 
2019-01-08T21:14:55.479+0000	ClassGraph	-- classfilePathWhiteBlackList: 
2019-01-08T21:14:55.479+0000	ClassGraph	-- classPackageWhiteBlackList: 
2019-01-08T21:14:55.479+0000	ClassGraph	-- classPackagePathWhiteBlackList: 
2019-01-08T21:14:55.479+0000	ClassGraph	-- moduleWhiteBlackList: 
2019-01-08T21:14:55.479+0000	ClassGraph	-- jarWhiteBlackList: 
2019-01-08T21:14:55.479+0000	ClassGraph	-- classpathElementResourcePathWhiteBlackList: 
2019-01-08T21:14:55.479+0000	ClassGraph	-- libOrExtJarWhiteBlackList: 
2019-01-08T21:14:55.479+0000	ClassGraph	-- performScan: true
2019-01-08T21:14:55.479+0000	ClassGraph	-- scanJars: true
2019-01-08T21:14:55.479+0000	ClassGraph	-- scanNestedJars: true
2019-01-08T21:14:55.480+0000	ClassGraph	-- scanDirs: true
2019-01-08T21:14:55.480+0000	ClassGraph	-- scanModules: true
2019-01-08T21:14:55.480+0000	ClassGraph	-- enableClassInfo: false
2019-01-08T21:14:55.480+0000	ClassGraph	-- enableFieldInfo: false
2019-01-08T21:14:55.480+0000	ClassGraph	-- enableMethodInfo: false
2019-01-08T21:14:55.480+0000	ClassGraph	-- enableAnnotationInfo: false
2019-01-08T21:14:55.480+0000	ClassGraph	-- enableStaticFinalFieldConstantInitializerValues: false
2019-01-08T21:14:55.481+0000	ClassGraph	-- enableInterClassDependencies: false
2019-01-08T21:14:55.481+0000	ClassGraph	-- enableExternalClasses: false
2019-01-08T21:14:55.481+0000	ClassGraph	-- enableSystemJarsAndModules: false
2019-01-08T21:14:55.481+0000	ClassGraph	-- ignoreClassVisibility: false
2019-01-08T21:14:55.481+0000	ClassGraph	-- ignoreFieldVisibility: false
2019-01-08T21:14:55.481+0000	ClassGraph	-- ignoreMethodVisibility: false
2019-01-08T21:14:55.481+0000	ClassGraph	-- disableRuntimeInvisibleAnnotations: false
2019-01-08T21:14:55.481+0000	ClassGraph	-- extendScanningUpwardsToExternalClasses: true
2019-01-08T21:14:55.481+0000	ClassGraph	-- enableRemoteJarScanning: false
2019-01-08T21:14:55.481+0000	ClassGraph	-- addedClassLoaders: null
2019-01-08T21:14:55.481+0000	ClassGraph	-- overrideClassLoaders: null
2019-01-08T21:14:55.482+0000	ClassGraph	-- addedModuleLayers: null
2019-01-08T21:14:55.482+0000	ClassGraph	-- overrideModuleLayers: null
2019-01-08T21:14:55.482+0000	ClassGraph	-- overrideClasspath: null
2019-01-08T21:14:55.482+0000	ClassGraph	-- classpathElementFilters: null
2019-01-08T21:14:55.482+0000	ClassGraph	-- initializeLoadedClasses: false
2019-01-08T21:14:55.482+0000	ClassGraph	-- removeTemporaryFilesAfterScan: false
2019-01-08T21:14:55.482+0000	ClassGraph	-- ignoreParentClassLoaders: false
2019-01-08T21:14:55.482+0000	ClassGraph	-- ignoreParentModuleLayers: false
2019-01-08T21:14:55.484+0000	ClassGraph	Finding classpath entries
2019-01-08T21:14:55.493+0000	ClassGraph	-- Finding classpath and modules
2019-01-08T21:14:55.494+0000	ClassGraph	---- System jars:
2019-01-08T21:14:55.494+0000	ClassGraph	------ Scanning disabled for rt.jar: /usr/lib/jvm/java-1.8.0-openjdk-1.8.0.191.b12-1.el7_6.x86_64/jre/lib/rt.jar
2019-01-08T21:14:55.494+0000	ClassGraph	------ Scanning disabled for lib or ext jar: /usr/lib/jvm/java-1.8.0-openjdk-1.8.0.191.b12-1.el7_6.x86_64/lib/dt.jar
2019-01-08T21:14:55.494+0000	ClassGraph	------ Scanning disabled for lib or ext jar: /usr/lib/jvm/java-1.8.0-openjdk-1.8.0.191.b12-1.el7_6.x86_64/lib/jconsole.jar
2019-01-08T21:14:55.494+0000	ClassGraph	------ Scanning disabled for lib or ext jar: /usr/lib/jvm/java-1.8.0-openjdk-1.8.0.191.b12-1.el7_6.x86_64/lib/sa-jdi.jar
2019-01-08T21:14:55.494+0000	ClassGraph	------ Scanning disabled for lib or ext jar: /usr/lib/jvm/java-1.8.0-openjdk-1.8.0.191.b12-1.el7_6.x86_64/lib/tools.jar
2019-01-08T21:14:55.495+0000	ClassGraph	------ Scanning disabled for lib or ext jar: /usr/lib/jvm/java-1.8.0-openjdk-1.8.0.191.b12-1.el7_6.x86_64/jre/lib/charsets.jar
2019-01-08T21:14:55.495+0000	ClassGraph	------ Scanning disabled for lib or ext jar: /usr/lib/jvm/java-1.8.0-openjdk-1.8.0.191.b12-1.el7_6.x86_64/jre/lib/jce.jar
2019-01-08T21:14:55.495+0000	ClassGraph	------ Scanning disabled for lib or ext jar: /usr/lib/jvm/java-1.8.0-openjdk-1.8.0.191.b12-1.el7_6.x86_64/jre/lib/jsse.jar
2019-01-08T21:14:55.495+0000	ClassGraph	------ Scanning disabled for lib or ext jar: /usr/lib/jvm/java-1.8.0-openjdk-1.8.0.191.b12-1.el7_6.x86_64/jre/lib/management-agent.jar
2019-01-08T21:14:55.495+0000	ClassGraph	------ Scanning disabled for lib or ext jar: /usr/lib/jvm/java-1.8.0-openjdk-1.8.0.191.b12-1.el7_6.x86_64/jre/lib/resources.jar
2019-01-08T21:14:55.495+0000	ClassGraph	------ Scanning disabled for lib or ext jar: /usr/lib/jvm/java-1.8.0-openjdk-1.8.0.191.b12-1.el7_6.x86_64/jre/lib/ext/cldrdata.jar
2019-01-08T21:14:55.495+0000	ClassGraph	------ Scanning disabled for lib or ext jar: /usr/lib/jvm/java-1.8.0-openjdk-1.8.0.191.b12-1.el7_6.x86_64/jre/lib/ext/dnsns.jar
2019-01-08T21:14:55.495+0000	ClassGraph	------ Scanning disabled for lib or ext jar: /usr/lib/jvm/java-1.8.0-openjdk-1.8.0.191.b12-1.el7_6.x86_64/jre/lib/ext/jaccess.jar
2019-01-08T21:14:55.495+0000	ClassGraph	------ Scanning disabled for lib or ext jar: /usr/lib/jvm/java-1.8.0-openjdk-1.8.0.191.b12-1.el7_6.x86_64/jre/lib/ext/localedata.jar
2019-01-08T21:14:55.496+0000	ClassGraph	------ Scanning disabled for lib or ext jar: /usr/lib/jvm/java-1.8.0-openjdk-1.8.0.191.b12-1.el7_6.x86_64/jre/lib/ext/nashorn.jar
2019-01-08T21:14:55.496+0000	ClassGraph	------ Scanning disabled for lib or ext jar: /usr/lib/jvm/java-1.8.0-openjdk-1.8.0.191.b12-1.el7_6.x86_64/jre/lib/ext/sunec.jar
2019-01-08T21:14:55.496+0000	ClassGraph	------ Scanning disabled for lib or ext jar: /usr/lib/jvm/java-1.8.0-openjdk-1.8.0.191.b12-1.el7_6.x86_64/jre/lib/ext/sunjce_provider.jar
2019-01-08T21:14:55.496+0000	ClassGraph	------ Scanning disabled for lib or ext jar: /usr/lib/jvm/java-1.8.0-openjdk-1.8.0.191.b12-1.el7_6.x86_64/jre/lib/ext/sunpkcs11.jar
2019-01-08T21:14:55.496+0000	ClassGraph	------ Scanning disabled for lib or ext jar: /usr/lib/jvm/java-1.8.0-openjdk-1.8.0.191.b12-1.el7_6.x86_64/jre/lib/ext/zipfs.jar
2019-01-08T21:14:55.496+0000	ClassGraph	------ Scanning disabled for lib or ext jar: /usr/java/packages/lib/ext/bcprov.jar
2019-01-08T21:14:55.506+0000	ClassGraph	---- Found ClassLoaders:
2019-01-08T21:14:55.507+0000	ClassGraph	------ sun.misc.Launcher$AppClassLoader@659e0bfd
2019-01-08T21:14:55.507+0000	ClassGraph	---- Found system modules:
2019-01-08T21:14:55.507+0000	ClassGraph	------ [None]
2019-01-08T21:14:55.507+0000	ClassGraph	---- Found non-system modules:
2019-01-08T21:14:55.507+0000	ClassGraph	------ [None]
2019-01-08T21:14:55.508+0000	ClassGraph	---- ClassLoaderHandlers:
2019-01-08T21:14:55.518+0000	ClassGraph	------ nonapi.io.github.classgraph.classloaderhandler.AntClassLoaderHandler
2019-01-08T21:14:55.518+0000	ClassGraph	------ nonapi.io.github.classgraph.classloaderhandler.EquinoxClassLoaderHandler
2019-01-08T21:14:55.518+0000	ClassGraph	------ nonapi.io.github.classgraph.classloaderhandler.EquinoxContextFinderClassLoaderHandler
2019-01-08T21:14:55.518+0000	ClassGraph	------ nonapi.io.github.classgraph.classloaderhandler.FelixClassLoaderHandler
2019-01-08T21:14:55.518+0000	ClassGraph	------ nonapi.io.github.classgraph.classloaderhandler.JBossClassLoaderHandler
2019-01-08T21:14:55.518+0000	ClassGraph	------ nonapi.io.github.classgraph.classloaderhandler.WeblogicClassLoaderHandler
2019-01-08T21:14:55.518+0000	ClassGraph	------ nonapi.io.github.classgraph.classloaderhandler.WebsphereLibertyClassLoaderHandler
2019-01-08T21:14:55.518+0000	ClassGraph	------ nonapi.io.github.classgraph.classloaderhandler.WebsphereTraditionalClassLoaderHandler
2019-01-08T21:14:55.518+0000	ClassGraph	------ nonapi.io.github.classgraph.classloaderhandler.OSGiDefaultClassLoaderHandler
2019-01-08T21:14:55.518+0000	ClassGraph	------ nonapi.io.github.classgraph.classloaderhandler.SpringBootRestartClassLoaderHandler
2019-01-08T21:14:55.519+0000	ClassGraph	------ nonapi.io.github.classgraph.classloaderhandler.ParentLastDelegationOrderTestClassLoaderHandler
2019-01-08T21:14:55.519+0000	ClassGraph	------ nonapi.io.github.classgraph.classloaderhandler.JPMSClassLoaderHandler
2019-01-08T21:14:55.519+0000	ClassGraph	------ nonapi.io.github.classgraph.classloaderhandler.URLClassLoaderHandler
2019-01-08T21:14:55.519+0000	ClassGraph	---- ClassLoader sun.misc.Launcher$AppClassLoader@659e0bfd will be handled by nonapi.io.github.classgraph.classloaderhandler.URLClassLoaderHandler@6895b812
2019-01-08T21:14:55.520+0000	ClassGraph	---- sun.misc.Launcher$AppClassLoader@659e0bfd delegates to parent sun.misc.Launcher$ExtClassLoader@5acf9800 with order PARENT_FIRST
2019-01-08T21:14:55.520+0000	ClassGraph	---- ClassLoader sun.misc.Launcher$ExtClassLoader@5acf9800 will be handled by nonapi.io.github.classgraph.classloaderhandler.URLClassLoaderHandler@5b1c54e8
2019-01-08T21:14:55.521+0000	ClassGraph	---- Finding classpath elements in ClassLoaders
2019-01-08T21:14:55.521+0000	ClassGraph	------ Finding classpath elements in ClassLoader sun.misc.Launcher$ExtClassLoader@5acf9800
2019-01-08T21:14:55.521+0000	ClassGraph	-------- Ignoring duplicate classpath element: file:/usr/lib/jvm/java-1.8.0-openjdk-1.8.0.191.b12-1.el7_6.x86_64/jre/lib/ext/cldrdata.jar -> /usr/lib/jvm/java-1.8.0-openjdk-1.8.0.191.b12-1.el7_6.x86_64/jre/lib/ext/cldrdata.jar
2019-01-08T21:14:55.521+0000	ClassGraph	-------- Ignoring duplicate classpath element: file:/usr/lib/jvm/java-1.8.0-openjdk-1.8.0.191.b12-1.el7_6.x86_64/jre/lib/ext/dnsns.jar -> /usr/lib/jvm/java-1.8.0-openjdk-1.8.0.191.b12-1.el7_6.x86_64/jre/lib/ext/dnsns.jar
2019-01-08T21:14:55.521+0000	ClassGraph	-------- Ignoring duplicate classpath element: file:/usr/lib/jvm/java-1.8.0-openjdk-1.8.0.191.b12-1.el7_6.x86_64/jre/lib/ext/jaccess.jar -> /usr/lib/jvm/java-1.8.0-openjdk-1.8.0.191.b12-1.el7_6.x86_64/jre/lib/ext/jaccess.jar
2019-01-08T21:14:55.521+0000	ClassGraph	-------- Ignoring duplicate classpath element: file:/usr/lib/jvm/java-1.8.0-openjdk-1.8.0.191.b12-1.el7_6.x86_64/jre/lib/ext/localedata.jar -> /usr/lib/jvm/java-1.8.0-openjdk-1.8.0.191.b12-1.el7_6.x86_64/jre/lib/ext/localedata.jar
2019-01-08T21:14:55.521+0000	ClassGraph	-------- Ignoring duplicate classpath element: file:/usr/lib/jvm/java-1.8.0-openjdk-1.8.0.191.b12-1.el7_6.x86_64/jre/lib/ext/nashorn.jar -> /usr/lib/jvm/java-1.8.0-openjdk-1.8.0.191.b12-1.el7_6.x86_64/jre/lib/ext/nashorn.jar
2019-01-08T21:14:55.521+0000	ClassGraph	-------- Ignoring duplicate classpath element: file:/usr/lib/jvm/java-1.8.0-openjdk-1.8.0.191.b12-1.el7_6.x86_64/jre/lib/ext/sunec.jar -> /usr/lib/jvm/java-1.8.0-openjdk-1.8.0.191.b12-1.el7_6.x86_64/jre/lib/ext/sunec.jar
2019-01-08T21:14:55.522+0000	ClassGraph	-------- Ignoring duplicate classpath element: file:/usr/lib/jvm/java-1.8.0-openjdk-1.8.0.191.b12-1.el7_6.x86_64/jre/lib/ext/sunjce_provider.jar -> /usr/lib/jvm/java-1.8.0-openjdk-1.8.0.191.b12-1.el7_6.x86_64/jre/lib/ext/sunjce_provider.jar
2019-01-08T21:14:55.522+0000	ClassGraph	-------- Ignoring duplicate classpath element: file:/usr/lib/jvm/java-1.8.0-openjdk-1.8.0.191.b12-1.el7_6.x86_64/jre/lib/ext/sunpkcs11.jar -> /usr/lib/jvm/java-1.8.0-openjdk-1.8.0.191.b12-1.el7_6.x86_64/jre/lib/ext/sunpkcs11.jar
2019-01-08T21:14:55.522+0000	ClassGraph	-------- Ignoring duplicate classpath element: file:/usr/lib/jvm/java-1.8.0-openjdk-1.8.0.191.b12-1.el7_6.x86_64/jre/lib/ext/zipfs.jar -> /usr/lib/jvm/java-1.8.0-openjdk-1.8.0.191.b12-1.el7_6.x86_64/jre/lib/ext/zipfs.jar
2019-01-08T21:14:55.522+0000	ClassGraph	-------- Ignoring duplicate classpath element: file:/usr/java/packages/lib/ext/bcprov.jar -> /usr/java/packages/lib/ext/bcprov.jar
2019-01-08T21:14:55.522+0000	ClassGraph	------ Finding classpath elements in ClassLoader sun.misc.Launcher$AppClassLoader@659e0bfd
2019-01-08T21:14:55.522+0000	ClassGraph	-------- Found classpath element: file:/home/chrisng/git/classgraph-duplicate-bug/target/classes/ -> /home/chrisng/git/classgraph-duplicate-bug/target/classes
2019-01-08T21:14:55.522+0000	ClassGraph	-------- Found classpath element: file:/home/chrisng/git/classgraph-duplicate-bug/target/deps/classgraph-4.6.12.jar -> /home/chrisng/git/classgraph-duplicate-bug/target/deps/classgraph-4.6.12.jar
2019-01-08T21:14:55.522+0000	ClassGraph	---- Getting classpath entries from java.class.path
2019-01-08T21:14:55.523+0000	ClassGraph	------ Ignoring duplicate classpath element: /home/chrisng/git/classgraph-duplicate-bug/target/classes
2019-01-08T21:14:55.523+0000	ClassGraph	------ Found classpath element: /home/chrisng/git/classgraph-duplicate-bug/src/main/java/../../../target/deps/classgraph-4.6.12.jar
2019-01-08T21:14:55.523+0000	ClassGraph	-- Reading jarfile directories and manifest files (took 0.052924 sec)
2019-01-08T21:14:55.531+0000	ClassGraph	---- Opening jarfile /home/chrisng/git/classgraph-duplicate-bug/target/deps/classgraph-4.6.12.jar
2019-01-08T21:14:55.531+0000	ClassGraph	---- Opening jarfile /home/chrisng/git/classgraph-duplicate-bug/target/deps/classgraph-4.6.12.jar
2019-01-08T21:14:55.559+0000	ClassGraph	------ This is a multi-release jar
2019-01-08T21:14:55.566+0000	ClassGraph	-- Final classpath element order:
2019-01-08T21:14:55.566+0000	ClassGraph	---- /home/chrisng/git/classgraph-duplicate-bug/target/classes
2019-01-08T21:14:55.566+0000	ClassGraph	---- /home/chrisng/git/classgraph-duplicate-bug/target/deps/classgraph-4.6.12.jar
2019-01-08T21:14:55.566+0000	ClassGraph	---- /home/chrisng/git/classgraph-duplicate-bug/target/deps/classgraph-4.6.12.jar
2019-01-08T21:14:55.567+0000	ClassGraph	-- Scanning filenames within classpath elements
2019-01-08T21:14:55.568+0000	ClassGraph	---- Scanning directory classpath element /home/chrisng/git/classgraph-duplicate-bug/target/classes (took 0.001513 sec)
2019-01-08T21:14:55.569+0000	ClassGraph	------ Scanning directory: /home/chrisng/git/classgraph-duplicate-bug/target/classes (took 0.000132 sec)
2019-01-08T21:14:55.568+0000	ClassGraph	---- Scanning jarfile classpath element /home/chrisng/git/classgraph-duplicate-bug/target/deps/classgraph-4.6.12.jar (took 0.002796 sec)
2019-01-08T21:14:55.570+0000	ClassGraph	------ Found whitelisted resource: META-INF/maven/io.github.classgraph/classgraph/pom.properties
2019-01-08T21:14:55.570+0000	ClassGraph	------ Found whitelisted resource: META-INF/maven/io.github.classgraph/classgraph/pom.xml
2019-01-08T21:14:55.568+0000	ClassGraph	---- Scanning jarfile classpath element /home/chrisng/git/classgraph-duplicate-bug/target/deps/classgraph-4.6.12.jar (took 0.002685 sec)
2019-01-08T21:14:55.570+0000	ClassGraph	------ Found whitelisted resource: META-INF/maven/io.github.classgraph/classgraph/pom.properties
2019-01-08T21:14:55.570+0000	ClassGraph	------ Found whitelisted resource: META-INF/maven/io.github.classgraph/classgraph/pom.xml
2019-01-08T21:14:55.576+0000	ClassGraph	Masking classfiles
2019-01-08T21:14:55.576+0000	ClassGraph	Classfile scanning is disabled
2019-01-08T21:14:55.580+0000	ClassGraph	Completed (took 0.087426 sec)

Jan 08, 2019 9:14:55 PM nonapi.io.github.classgraph.utils.LogNode flush
INFO: 
Jan 08, 2019 9:14:55 PM nonapi.io.github.classgraph.utils.LogNode flush
INFO: 
Exception in thread "main" java.lang.AssertionError: Found duplicate path!
	at org.facboy.DuplicateClasspathBug.main(DuplicateClasspathBug.java:18)
```

**Comment by lukehutch** [comment]
I tried running from the commandline and from Eclipse, but can't replicate for some reason. Nevertheless, I added canonicalization to git master. Can you please test and let me know if it fixes the issue?

**Comment by lukehutch** [comment]
@facboy I tried running from the commandline and from Eclipse, but can't replicate for some reason. Nevertheless, I added canonicalization to git master. Can you please test and let me know if it fixes the issue?

**Comment by facboy** [comment]
Yes that fixed my test case (and also the original project where I encountered the problem).  Strange that you couldn't reproduce it :(.

**Comment by lukehutch** [comment]
OK, thanks for testing. This will be in the next release (I'm just waiting for feedback on one other issue first).

**Comment by lukehutch** [comment]
Released in 4.6.13. @facboy thanks for the report and testcase!
